presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -149,14 +151,21 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
                 throw new PrestoException(NO_NODES_AVAILABLE, "No nodes available to run query");
             }
 
-            InternalNode chosenNode = chooseLeastBusyNode(candidateNodes, assignmentStats::getTotalSplitCount, preferredNodeCount, maxSplitsPerNode);
-            if (chosenNode == null) {
+            NodeResult chosenNode = chooseLeastBusyNode(candidateNodes, assignmentStats::getTotalSplitCount, preferredNodeCount, maxSplitsPerNode);
+            if (!chosenNode.getNode().isPresent()) {
                 chosenNode = chooseLeastBusyNode(candidateNodes, assignmentStats::getQueuedSplitCountForStage, preferredNodeCount, maxPendingSplitsPerTask);
             }
 
-            if (chosenNode != null) {
-                assignment.put(chosenNode, split);
-                assignmentStats.addAssignedSplit(chosenNode);
+            if (chosenNode.getNode().isPresent()) {
+                split = new Split(
+                        split.getConnectorId(),
+                        split.getTransactionHandle(),
+                        split.getConnectorSplit(),
+                        split.getLifespan(),
+                        new SplitContext(chosenNode.ifShouldCache()));
```

Comment:
I remembered @wenleix once mentioned that changing `Split` in scheduling time might not be respected by Presto-on-Spark. I'll let @wenleix chime in here but would like to call it out so we are aware :)

Reply:
@wenleix mentioned to me as well. I vaguely remember this won't affect spark. Will let @wenleix comment.